libpromises/evalfunction.c

```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
-        if (sscanf(ip->classes, "%lf", &entrytime) != 1)
+        HostData *hd = SeqAt(host_data, i);
+        if (sscanf(hd->lastseen, "%lf", &entrytime) != 1)
         {
             Log(LOG_LEVEL_ERR, "Could not get host entry age");
             continue;
         }
 
-        if (return_address)
-        {
-            StringCopy(ip->name, address, sizeof(address));
-        }
-        else
+        switch (return_what)
         {
-            char hostname[NI_MAXHOST];
-            if (IPString2Hostname(hostname, ip->name, sizeof(hostname)) != -1)
-            {
-                StringCopy(hostname, address, sizeof(address));
-            }
-            else
+            case NAME:
             {
-                /* Not numeric address was requested, but IP was unresolvable. */
-                StringCopy(ip->name, address, sizeof(address));
+                char hostname[NI_MAXHOST];
+                if (IPString2Hostname(hostname, hd->address, sizeof(hostname)) != -1)
+                {
+                    StringCopy(hostname, ret_host_data, sizeof(ret_host_data));
+                }
+                else
+                {
+                    /* Not numeric address was requested, but IP was unresolvable. */
+                    StringCopy(hd->address, ret_host_data, sizeof(ret_host_data));
+                }
+                break;
             }
+            case ADDRESS:
+                StringCopy(hd->address, ret_host_data, sizeof(ret_host_data));
+                break;
+            case KEY:
+                StringCopy(hd->hostkey, ret_host_data, sizeof(ret_host_data));
+                break;
+            default:
+                ProgrammingError("Parser allowed invalid hostsseen() format argument");
         }
 
         if (entrytime < now - horizon)
         {
             Log(LOG_LEVEL_DEBUG, "Old entry");
 
-            if (RlistKeyIn(recent, address))
+            if (RlistKeyIn(recent, ret_host_data))
             {
-                Log(LOG_LEVEL_DEBUG, "There is recent entry for this address. Do nothing.");
+                Log(LOG_LEVEL_DEBUG, "There is recent entry for this ret_host_data. Do nothing.");
             }
             else
             {
                 Log(LOG_LEVEL_DEBUG, "Adding to list of aged hosts.");
-                RlistPrependScalarIdemp(&aged, address);
+                RlistPrependScalarIdemp(&aged, ret_host_data);
             }
         }
         else
         {
             Log(LOG_LEVEL_DEBUG, "Recent entry");
 
-            Rlist *r = RlistKeyIn(aged, address);
+            Rlist *r = RlistKeyIn(aged, ret_host_data);
             if (r)
             {
                 Log(LOG_LEVEL_DEBUG, "Purging from list of aged hosts.");
                 RlistDestroyEntry(&aged, r);
             }
 
             Log(LOG_LEVEL_DEBUG, "Adding to list of recent hosts.");
-            RlistPrependScalarIdemp(&recent, address);
+            RlistPrependScalarIdemp(&recent, ret_host_data);
         }
```

Comment:
As an optimization, you could do the filtering first, before rendering `ret_host_data`. But let's wait with this and do it in a follow-up PR once you have everything working.

Reply:
i was abit skeptical about this algorithm but true its not really relevant the current PR